File: Autumn/task7/task7/List.c
#include <stdio.h>
#include <stdlib.h>
#include "List.h"



int contain(list *first, int keyForSearch)
{
	if (first == NULL)
	{
		return 0;
	}
	
	list *temp = first;
	while (temp->key != keyForSearch)
	{
		if (temp->next != NULL)
		{
			temp = temp->next;
		}
		else
		{
			return 0;
		}
	}
	return 1;
}

int addFirstElement(list **first, int key, int value)
{
	if (contain(*first, key))
	{
		return 0;
	}

	list *newFirst = malloc(sizeof(list));
	newFirst->next = *first;
	newFirst->value = value;
	newFirst->key = key;

	*first = newFirst;
	return 1;
}

void deleteFirst(list **first)
{
	if (*first == NULL)
	{
		return;
	}

	list *oldFirst = *first;
	*first = (*first)->next;
	free(oldFirst);
}

int deleteElement(list **first, int key)
{
	if (*first == NULL)
	{
		return 0;
	}

	if (!contain(*first, key))
	{
		return 0;
	}

	if ((*first)->key == key)
	{
		deleteFirst(*&first);
		return 1;
	}

	list *temp = *first;
	while ((temp->next != NULL) && (temp->next->key != key))
	{
		temp = temp->next;
	}
	if (temp->next == NULL)
	{
		return 0;
	}
	list *oldElement = temp->next;
	list *elementNext = temp->next;
	temp->next = elementNext->next;
	free(oldElement);

	return 1;
}

int getValueByKey(list *first, int key, int *correct)
{
	if (first == NULL)
	{
		*correct = 0;
		return 0;
	}
	
	list *temp = first;

	while (temp->key != key)
	{
		if (temp->next != NULL)
		{
			temp = temp->next;
		}
		else
		{
			*correct = 0;
			return 0;
		}
	}
	*correct = 1;
	return temp->value;
}

void deleteList(list **first)
{
	while (*first != NULL)
	{
		deleteFirst(*&first);
	}
}

void printList(list *first)
{
	if (first == NULL)
	{
		return;
	}

	list *temp = first;
	while (temp->next != NULL)
	{
		printf("%d", temp->key);
		putchar('-');
		printf("%d", temp->value);
		putchar(' ');
		temp = temp->next;
	}
	printf("%d", temp->key);
	putchar('-');
	printf("%d", temp->value);
	putchar('\n');
}



void addElementsForTest(list **list)
{
	addFirstElement(*&list, 1, 11);
	addFirstElement(*&list, 2, 12);
	addFirstElement(*&list, 3, 13);
	addFirstElement(*&list, 4, 14);
	addFirstElement(*&list, 5, 15);
}

int testAddElement()
{
	list *list = NULL;
	addElementsForTest(&list);

	if (contain(list, 1) && contain(list, 2) && contain(list, 3) && contain(list, 4) && contain(list, 5))
	{
		deleteList(&list);
		return 1;
	}
	deleteList(&list);
	return 0;
}

int testContain()
{
	list *list = NULL;
	addElementsForTest(&list);

	if (contain(list, 1) && contain(list, 3) && contain(list, 5) && (!contain(list, 6)) && (!contain(list, 7)))
	{
		deleteList(&list);
		return 1;
	}
	deleteList(&list);
	return 0;
}

int testDeleteElement()
{
	list *list = NULL;
	addElementsForTest(&list);

	if (deleteElement(&list, 6))
	{
		deleteList(&list);
		return 0;
	}

	if ((!deleteElement(&list, 5)) || (!deleteElement(&list, 3)) || (!deleteElement(&list, 1)))
	{
		deleteList(&list);
		return 0;
	}

	if (contain(list, 1) || contain(list, 3) || contain(list, 1))
	{
		deleteList(&list);
		return 0;
	}
	deleteList(&list);
	return 1;
}

int getValueByKeyTest()
{
	list *list = NULL;
	addElementsForTest(&list);
	int correct = -1;

	if ((getValueByKey(list, 2, &correct) != 12) || (correct != 1))
	{
		deleteList(&list);
		return 0;
	}

	if ((getValueByKey(list, 6, &correct) != 0) || (correct != 0))
	{
		deleteList(&list);
		return 0;
	}

	deleteList(&list);
	return 1;
}

int testList()
{
	return (testAddElement() && testContain() && testDeleteElement() && getValueByKeyTest());
}

File: Spring/task4.HashTable/task4.HashTable/List.cs
using System;

namespace task4.HashTable
{
    public class List<T>
    {
        private class ListElement
        {
            private T value;

            public T Value
            {
                get
                {
                    return value;
                }
                private set
                {
                    this.value = value;
                }
            }

            private int key;

            public int Key
            {
                get
                {
                    return key;
                }
                private set
                {
                    key = value;
                }
            }

            private ListElement next;

            public ListElement Next
            {
                get
                {
                    return next;
                }
                set
                {
                    next = value;
                }
            }

            public ListElement(T value, int key, ListElement next)
            {
                this.Value = value;
                this.Key = key;
                this.Next = next;
            }
        }

        private int length;

        public int Length
        {
            get
            {
                return length;
            }
            private set
            {
                length = value;
            }
        }

        private ListElement first;

        public List()
        {
            this.first = null;
            this.Length = 0;
        }

        public void AddFirst(T value, int key)
        {
            var newElement = new ListElement(value, key, first);
            first = newElement;
            Length++;
        }

        public T GetValueByKey(int key)
        {
            ListElement temp = first;
            while(temp != null)
            {
                if (temp.Key == key)
                {
                    return temp.Value;
                }
                temp = temp.Next;
            }
            return default(T);
        }

        public void DeleteElement(int key)
        {
            while ((first != null) && (first.Key == key))
            {
                first = first.Next;
                Length--;
            }

            ListElement temp = first;
            while (temp != null)
            {
                while ((temp.Next != null) && (temp.Next.Key == key))
                {
                    temp.Next = temp.Next.Next;
                    Length--;
                }
                temp = temp.Next;
            }
        }

        public bool Contains(int key)
        {
            ListElement temp = first;
            while (temp != null)
            {
                if (temp.Key == key)
                {
                    return true;
                }
                temp = temp.Next;
            }
            return false;
        }

        public void PrintList()
        {
            if(first == null)
            {
                return;
            }
            ListElement temp = first;
            while (temp != null)
            {
                Console.Write(temp.Value);
                Console.Write(" ");
                temp = temp.Next;
            }
            Console.WriteLine();
        }

        public void DeleteList()
        {
            first = null;
        }
    }
}
